spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java

```diff
@@ -80,42 +85,77 @@ public KStreamBinder(KafkaBinderConfigurationProperties binderConfigurationPrope
 	@SuppressWarnings("unchecked")
 	protected Binding<KStream<Object, Object>> doBindProducer(String name, KStream<Object, Object> outboundBindTarget,
 															ExtendedProducerProperties<KStreamProducerProperties> properties) {
-		ExtendedProducerProperties<KafkaProducerProperties> extendedProducerProperties = new ExtendedProducerProperties<KafkaProducerProperties>(
+		ExtendedProducerProperties<KafkaProducerProperties> extendedProducerProperties = new ExtendedProducerProperties<>(
 				new KafkaProducerProperties());
 		this.kafkaTopicProvisioner.provisionProducerDestination(name, extendedProducerProperties);
-		outboundBindTarget = outboundBindTarget
-				.map((k, v) -> KeyValue.pair(k, ((Message<Object>) v).getPayload()));
 
-		Serde<?> keySerde = Serdes.ByteArray();
-		Serde<?> valueSerde = Serdes.ByteArray();
-		if (properties.isUseNativeEncoding()) {
-			outboundBindTarget.to(name, Produced.with((Serde<Object>) keySerde, (Serde<Object>) valueSerde));
-		}
-		else {
-			try {
-				if (StringUtils.hasText(properties.getExtension().getKeySerde())) {
-					keySerde = Utils.newInstance(properties.getExtension().getKeySerde(), Serde.class);
-					if (keySerde instanceof Configurable) {
-						((Configurable) keySerde).configure(streamsConfig.originals());
-					}
-				}
-				else {
-					keySerde = this.binderConfigurationProperties.getConfiguration().containsKey("key.serde") ?
-							Utils.newInstance(this.binderConfigurationProperties.getConfiguration().get("key.serde"), Serde.class) : Serdes.ByteArray();
+		Serde<?> keySerde = getKeySerde(properties);
+		Serde<?> valueSerde = getValueSerde(properties);
+
+		to(properties.isUseNativeEncoding(), name, outboundBindTarget, (Serde<Object>) keySerde, (Serde<Object>) valueSerde);
+
+		return new DefaultBinding<>(name, null, outboundBindTarget, null);
+	}
+
+	private Serde<?> getKeySerde(ExtendedProducerProperties<KStreamProducerProperties> properties) {
+		Serde<?> keySerde;
+		try {
+			if (StringUtils.hasText(properties.getExtension().getKeySerde())) {
+				keySerde = Utils.newInstance(properties.getExtension().getKeySerde(), Serde.class);
+				if (keySerde instanceof Configurable) {
+					((Configurable) keySerde).configure(streamsConfig.originals());
 				}
+			}
+			else {
+				keySerde = this.binderConfigurationProperties.getConfiguration().containsKey("key.serde") ?
+						Utils.newInstance(this.binderConfigurationProperties.getConfiguration().get("key.serde"), Serde.class) : Serdes.ByteArray();
+			}
```

Comment:
Can you raise an issue to revisit this? I am not fully grasping context here yet, but I just made similar change to our partitioning. Basically I would love to avoid `newInstance` in favor of making things Spring configured. It was easy for partitioning and should be easy here as well, but we can talk about it as separate improvement.

Reply:
Issue created: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/288